### ISSUE TITLE
questions for dpo with lora

### ISSUE THREAD

**Issue by Gucci233** [2U1/Qwen2-VL-Finetune#159]
Hello, the original DPO script was designed for full-parameter fine-tuning, but I modified it to use LoRA fine-tuning instead, and used a simple dataset (by adding “dear” and “damn” respectively). During training, the loss quickly dropped to zero, indicating that the model had learned something.

<img width="990" height="328" alt="Image" src="https://github.com/user-attachments/assets/1136f6c6-7a7c-4ff1-93f6-9736f041f8de" />

However, after I merged the LoRA parameters into the base model and tested it, the outputs did not contain "dear" or "damn" as expected.

I'm not sure whether the issue lies in the training parameters or something went wrong during the merging process. Below are my training script and config file.
dpo_script.sh
`
deepspeed src/train/train_sft.py \
    --use_liger True \
    --lora_enable True \
    --use_dora False \
    --lora_namespan_exclude "['lm_head', 'embed_tokens']" \
    --lora_rank 64 \
    --lora_alpha 64 \
    --lora_dropout 0.05 \
    --num_lora_modules -1 \
    --deepspeed scripts/zero3.json \
    --model_id $MODEL_NAME \
    --data_path /home/cvpr/member/gqw/llm_pr/Qwen2-VL-Finetune/data/train.json \
    --image_folder /home/cvpr/member/gqw/llm_pr/textvqa_grounding_task_qwen2.5-vl-ft/data/train \
    --remove_unused_columns False \
    --freeze_vision_tower True \
    --freeze_llm True \
    --freeze_merger False \
    --bf16 True \
    --fp16 False \
    --disable_flash_attn2 False \
    --output_dir output/testing_lora \
    --num_train_epochs 1 \
    --per_device_train_batch_size $BATCH_PER_DEVICE \
    --gradient_accumulation_steps $GRAD_ACCUM_STEPS \
    --image_min_pixels $((256 * 28 * 28)) \
    --image_max_pixels $((1280 * 28 * 28)) \
    --learning_rate 1e-4 \
    --merger_lr 1e-5 \
    --vision_lr 2e-6 \
    --weight_decay 0.1 \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --gradient_checkpointing True \
    --report_to tensorboard \
    --lazy_preprocess True \
    --save_strategy "steps" \
    --save_steps 100 \
    --save_total_limit 10 \
    --dataloader_num_workers 4`

config:
`{
  "_attn_implementation_autoset": true,
  "architectures": [
    "Qwen2_5_VLForConditionalGeneration"
  ],
  "attention_dropout": 0.0,
  "bos_token_id": 151643,
  "eos_token_id": 151645,
  "hidden_act": "silu",
  "hidden_size": 3584,
  "image_token_id": 151655,
  "initializer_range": 0.02,
  "intermediate_size": 18944,
  "max_position_embeddings": 128000,
  "max_window_layers": 28,
  "model_type": "qwen2_5_vl",
  "num_attention_heads": 28,
  "num_hidden_layers": 28,
  "num_key_value_heads": 4,
  "rms_norm_eps": 1e-06,
  "rope_scaling": {
    "mrope_section": [
      16,
      24,
      24
    ],
    "rope_type": "default",
    "type": "default"
  },
  "rope_theta": 1000000.0,
  "sliding_window": 32768,
  "tie_word_embeddings": false,
  "torch_dtype": "bfloat16",
  "transformers_version": "4.51.3",
  "use_cache": true,
  "use_sliding_window": false,
  "video_token_id": 151656,
  "vision_config": {
    "depth": 32,
    "fullatt_block_indexes": [
      7,
      15,
      23,
      31
    ],
    "hidden_act": "silu",
    "hidden_size": 1280,
    "in_channels": 3,
    "in_chans": 3,
    "intermediate_size": 3420,
    "model_type": "qwen2_5_vl",
    "num_heads": 16,
    "out_hidden_size": 3584,
    "patch_size": 14,
    "spatial_merge_size": 2,
    "spatial_patch_size": 14,
    "temporal_patch_size": 2,
    "tokens_per_second": 2,
    "torch_dtype": "bfloat16",
    "window_size": 112
  },
  "vision_end_token_id": 151653,
  "vision_start_token_id": 151652,
  "vision_token_id": 151654,
  "vocab_size": 152064
}
`

**Comment by 2U1** [comment]
### Why the loss instantly drops to 0
* `ref_model` is **None**, so `DPOTrainer` clones the *same base weights* (LoRA removed) as the reference → policy ≈ reference → `chosen_logp-rejected_logp ≈ 0` → loss → 0.
* `lm_head` and `embed_tokens` are excluded from LoRA, limiting any change in token probabilities.
* Very small dataset + few training steps makes the margin saturate quickly.

### Quick fix

These are the options that you could try.

1. **Load an explicit reference model**  
```python
   ref_model = AutoModelForCausalLM.from_pretrained(
       base_id, load_in_4bit=True, device_map={"": "cuda"}
   )
   ref_model.eval();  ref_model.requires_grad_(False)
```

2. Include lm_head and embed_tokens in LoRA (or set bias="all").

3. Train for a few hundred update steps; log chosen_logp - rejected_logp along with the loss.

**Comment by Gucci233** [comment]
Thank you for your suggestion. I conducted some experiments and encountered a few issues:

When I trained using the 7B model and tested with the weights from 100 epochs, I manually mounted the LoRA to the model and ran direct inference. In this setup, the output already included the word "damn", but there were repetitions in the later part of the response. However, after merging the parameters and using the visualization tool you provided, the output appeared normal.

<img width="315" height="366" alt="Image" src="https://github.com/user-attachments/assets/664304d3-b4a3-4a9c-bc8f-c27c8710bad5" />

<img width="2276" height="756" alt="Image" src="https://github.com/user-attachments/assets/c686b7f9-9603-4632-b981-c378603d7cfa" />

I'm wondering if this discrepancy is caused by the temperature parameter or the sampling strategy, which may be leading to inconsistent outputs. As follow is my test code.
```
import os
import torch
from peft import PeftModel
from transformers import AutoProcessor, Qwen2_5_VLForConditionalGeneration
from PIL import Image

model_path = "/home/cvpr/member/gqw/llm_pr/Qwen2-VL-Finetune/output/rlaif_dpo_tenth_lora/checkpoint-100"  # LoRA权重和非LoRA权重的目录
base_model_path = "/home/cvpr/member/gqw/llm_pr/Qwen2.5-VL-7B-Instruct"
image_path = "/home/cvpr/member/gqw/llm_pr/Qwen2-VL-Finetune/data/TextVQA/train_images/8733d0a1351be922.jpg"
device = "cuda" if torch.cuda.is_available() else "cpu"


model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    base_model_path,
    low_cpu_mem_usage=True,
    torch_dtype=torch.float16 if torch.cuda.is_available() else torch.float32
)
model.to(device)


non_lora_path = os.path.join(model_path, 'non_lora_state_dict.bin')
non_lora_trainables = torch.load(non_lora_path, map_location='cpu')


non_lora_trainables = {
    (k[11:] if k.startswith('base_model.') else k): v
    for k, v in non_lora_trainables.items()
}
if any(k.startswith('model.model.') for k in non_lora_trainables):
    non_lora_trainables = {
        (k[6:] if k.startswith('model.') else k): v
        for k, v in non_lora_trainables.items()
    }


model.load_state_dict(non_lora_trainables, strict=False)


model = PeftModel.from_pretrained(model, model_path,is_trainable=False)

model.eval()


processor = AutoProcessor.from_pretrained(base_model_path, trust_remote_code=True)


prompt = "Describe this image."
image = Image.open(image_path).convert("RGB")


messages = [
    {
        "role": "user",
        "content": [
            {"type": "image"},
            {"type": "text", "text": prompt}
        ]
    }
]


text = processor.apply_chat_template(messages, add_generation_prompt=True, tokenize=False)
inputs = processor(text=[text], images=[image], return_tensors="pt")


inputs = {k: v.to(device) for k, v in inputs.items()}


print("\n--- begin ---")
print(f"input: {prompt}")

with torch.no_grad():
    generated_ids = model.generate(**inputs, max_new_tokens=512, do_sample=False)

response_ids = generated_ids[:, inputs['input_ids'].shape[1]:]
output = processor.batch_decode(response_ids, skip_special_tokens=True)[0]


print("\n📝 output:", output)
print("--- finish ---\n")
```

Another issue is that when I trained a smaller 3B model, repetitive outputs started appearing as early as 100 epochs. What could be the reason behind this?

Also, why does the model seem to lose the ability to output words like "damn" as training progresses such as 1000epoch?

Looking forward to your reply.

**Comment by 2U1** [comment]
- Capacity / entropy ceiling Smaller models hit sharp logits sooner → greedy decoding gets stuck in self-loops unless you add temperature or repetition_penalty.

- Over-optimisation Training for 1 000 epochs on a micro-dataset drives the LoRA A/B matrices toward a rank-1 solution that satisfies the ranking loss but drifts back toward the base distribution elsewhere—so “damn” may no longer be the arg-max even though the ranking objective is still 100 % accurate.

- Weight-decay on LoRA With weight_decay=0.1 the A/B weights slowly shrink; after enough epochs the effective delta becomes tiny. Lower weight_decay (or turn it off for LoRA layers) if you really need very long training.


Also, how many samples are you training?

**Comment by Gucci233** [comment]
Thank you for your suggestions and reply. After some testing, I’ve understood the issue. Thanks again!
